### ISSUE TITLE
optimizer.zero_grad out of order

### ISSUE THREAD

**Issue by jsta** [lanl/Yoke#59]
I believe the following chunk that appears throughout [src/yoke/torch_training_utils.py](src/yoke/torch_training_utils.py):

```python
optimizer.zero_grad(set_to_none=True)  # Possible speed-up
loss.mean().backward()
optimizer.step()
```

should instead be:

```python
loss.mean().backward()
optimizer.step()
optimizer.zero_grad(set_to_none=True)  # Possible speed-up
```

as is, I think there is gradient leakage across epochs and higher than necessary usage overall:

with order corrected:
![Image](https://github.com/user-attachments/assets/ef96d0b6-eb13-4bb1-a998-fdc8263c7070)

without order correction:
![Image](https://github.com/user-attachments/assets/f3bd585d-3574-4c4e-9f3e-09f1bc57678a)

**Comment by jsta** [comment]
code is https://github.com/jsta/pytorch_memory/blob/main/create_profiler.py
